### PR TITLE
Make ScopedFingerprintWriter properly reenterable

### LINKED ISSUE #30925: Configuration cache fails to handle `ValueSource` with undeclared input parameters
## Expected Behavior

It should be possible to evaluate a `ValueSource` at configuration time, even when it has parameters that resolve to undeclared inputs .

## Actual Behavior
When the parameters of a `ValueSource` include undeclared inputs, evaluating it at configuration time fails with a cryptic error message:

```
* What went wrong:
A problem occurred configuring root project 'rsv52'.
> Could not create task ':build'.
   > Failed to notify value listener.
      > Configuration cache state could not be cached: field `__string__` of `CustomValueSource$Parameters_Decorated` bean found in field `parameters` of `org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$DefaultObtainedValue` bean found in field `obtainedValue` of `org.gradle.internal.cc.impl.fingerprint.ConfigurationCacheFingerprint$ValueSource` bean found in build file 'build.gradle': error writing value of type 'org.gradle.api.internal.provider.DefaultProperty'
         > Coroutine didn't run to completion.
```

## Reproducer

```groovy
    def 'reentrant fingerprint'() {
        given:
        buildFile """
            abstract class CustomValueSource implements ValueSource<String, Parameters> {
                interface Parameters extends ValueSourceParameters {
                    Property<String> getString()
                }

                @Override String obtain() {
                    return parameters.string.get()
                }
            }

            abstract class PrintTask extends DefaultTask {
                @Input abstract Property<String> getMessage()
                @TaskAction def doIt() {
                    println message.get()
                }
            }

            tasks.register('build', PrintTask) {
                message = providers.of(CustomValueSource) {
                    parameters {
                        string = provider {
                            System.getProperty('MY_SYSTEM_PROPERTY', '42')
                        }
                    }
                }.get() // explicitly calling get to trigger reentrant fingerprint behavior

            }
        """

        when:
        configurationCacheRun 'build'
        configurationCacheRun 'build'

        then:
        outputContains '42'
    }
```

### PR DESCRIPTION
**Pull request by mlopatkin**
It is possible that writing a CC input into the fingerprint triggers another input to be written. Writing synchronously is incorrect because it breaks the layout of the enclosing input and causes a deserialization error.

This commit collects all nested writes (which happen on the same thread) and writes them after completing the enclosing one, so the layout of the fingerprint file - a sequence of inputs - is preserved.

Fixes #30925

### CODE REVIEW

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt [hunk 1]
```diff
@@ -21,38 +21,78 @@ import org.gradle.internal.serialize.graph.CloseableWriteContext
...
     }
 
     fun write(value: T, trace: PropertyTrace? = null) {
         synchronized(writeContext) {
-            withPropertyTrace(trace) {
-                unsafeWrite(value)
+            if (isWriting) {
+                // Writing a value may cause another value to be written, e.g. when `provider {}` used as a ValueSource parameter reads an environment variable.
+                // These are configuration inputs, but we cannot write them immediately, or we will break the layout of the outer value.
+                // It doesn't expect that an Input comes right in a middle of a Provider's serialized form.
```
Comment (abstratt) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt:55]: These comments are helpful, but I can't follow this specific sentence. Maybe rephrase it?

**Review by abstratt — APPROVED** [review]

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt (hunk 1 above)
Comment (mlopatkin) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt]: > These comments are helpful, but I can't follow this specific sentence. Maybe rephrase it?

Thanks, I've elaborated the description.

**Comment by mlopatkin** [comment]
@bot-gradle test ACC

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt [hunk 2]
```diff
@@ -21,38 +21,82 @@ import org.gradle.internal.serialize.graph.CloseableWriteContext
...
     override fun close() {
         // we synchronize access to all resources used by callbacks
         // in case there was still an event being dispatched at closing time.
         synchronized(writeContext) {
-            unsafeWrite(null)
+            require(!isWriting && pendingWrites.isEmpty()) {
+                "Incomplete write detected"
+            }
+
+            writeValueLocked(null)
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt:45]: 🤔 This `Locked` suffix reads backward to me. `writeValueLocked` makes me think the method will acquire the lock to write value.

Comment (mlopatkin) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt:45]: > 🤔 This `Locked` suffix reads backward to me. `writeValueLocked` makes me think the method will acquire the lock to write value.

I've picked this convention when working with AOSP, but I see how it can be confusing.

Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ScopedFingerprintWriter.kt]: I'd still use some suffix or prefix to differentiate this method from the synchronizing one but we can come up with something later.

**Review by bamboo — APPROVED** [review]
